icelake/src/io/mod.rs

```diff
@@ -17,22 +17,214 @@ pub use scan::*;
 use arrow_array::RecordBatch;
 use arrow_schema::SchemaRef;
 
-use crate::types::DataFileBuilder;
+use crate::types::{DataFileBuilder, StructValue};
 use crate::Result;
+pub mod input_wrapper;
+pub use input_wrapper::*;
+
+type DefaultInput = RecordBatch;
 
 #[async_trait::async_trait]
-pub trait RecordBatchWriter: Send + 'static {
-    async fn write(&mut self, batch: RecordBatch) -> Result<()>;
-    async fn close(&mut self) -> Result<Vec<DataFileBuilder>>;
+pub trait FileWriterBuilder: Send + Sync + Clone + 'static {
+    type R: FileWriter;
+    async fn build(self, schema: &SchemaRef, file_name: &str) -> Result<Self::R>;
+}
+
+#[async_trait::async_trait]
+pub trait FileWriter: Send + 'static {
+    type R: FileWriteResult;
+    async fn write(&mut self, batch: &RecordBatch) -> Result<()>;
+    async fn close(self) -> Result<Self::R>;
+}
+
+pub trait FileWriteResult: Send + 'static {
+    type R: IcebergWriteResult;
+    /// return None Indicates the result is empty.
+    fn to_iceberg_result(self) -> Option<Self::R>;
 }
 
 #[async_trait::async_trait]
-pub trait RecordBatchWriterBuilder: Send + Sync + Clone + 'static {
-    type R: RecordBatchWriter;
+pub trait IcebergWriterBuilder: Send + Clone + 'static {
+    type R;
     async fn build(self, schema: &SchemaRef) -> Result<Self::R>;
 }
 
-pub trait SingletonWriter: RecordBatchWriter {
+#[async_trait::async_trait]
+pub trait IcebergWriter<I = DefaultInput>: Send + 'static {
+    type R: IcebergWriteResult;
+    async fn write(&mut self, input: I) -> Result<()>;
+    async fn flush(&mut self) -> Result<Self::R>;
+}
+
+pub trait IcebergWriteResult: Send + Sync + 'static + Default {
+    fn with_file_path(&mut self, file_name: String) -> &mut Self;
+    fn with_content(&mut self, content: crate::types::DataContentType) -> &mut Self;
+    fn with_equality_ids(&mut self, equality_ids: Vec<i32>) -> &mut Self;
+    fn with_partition(&mut self, partition_value: Option<StructValue>) -> &mut Self;
+    fn combine(&mut self, other: Self);
+    fn flush(&mut self) -> Self;
+}
+
+pub trait SingletonWriter {
```

Comment:
Since the metrics all starts with `current_`? Is it possble make it part of `FileWriter`?

Reply:
It's possible to make `row_num` and `writen_size` in FileWriter and IcebergWriter. But `current_file` may not be appropriate, how about a writer write to multiple files in current such as PartitionWriter. That's why we need the trait `SingletonWriter`.

Reply:
Oh, I see. But the name `SingletonWriter` looks weird, how about renaming it to `SingleFileWriter`?